demos/coreMQTT/mqtt_demo_serializer.c

```diff
@@ -0,0 +1,1620 @@
+            LogInfo( ( "Short delay before starting the next iteration.... \r\n\r\n" ) );
+            vTaskDelay( mqttexampleDELAY_BETWEEN_DEMO_ITERATIONS );
+        }
+        else
+        {
+            LogInfo( ( "RunCoreMqttSerializerDemo() failed an iteration. "
+                       "Total free heap is %u.\r\n",
+                       xPortGetFreeHeapSize() ) );
+            break;
+        }
+    }
+
+    return ( xStatus == pdPASS ) ? EXIT_SUCCESS : EXIT_FAILURE;
+}
+/*-----------------------------------------------------------*/
+
+static void prvGracefulShutDown( Socket_t xSocket )
+{
+    uint8_t ucDummy[ 20 ];
+    const TickType_t xShortDelay = pdMS_TO_TICKS( 250 );
+    BaseType_t xShutdownLoopCount = 0;
+
+    if( xSocket != ( Socket_t ) 0 )
+    {
+        if( xSocket != SOCKETS_INVALID_SOCKET )
+        {
+            /* Initiate graceful shutdown. */
+            SOCKETS_Shutdown( xSocket, SOCKETS_SHUT_RDWR );
+
+            /* Wait for the socket to disconnect gracefully (indicated by FreeRTOS_recv()
+             * returning a FREERTOS_EINVAL error) before closing the socket. */
+            while( SOCKETS_Recv( xSocket, ucDummy, sizeof( ucDummy ), 0 ) >= 0 )
+            {
+                /* Wait for shutdown to complete.  If a receive block time is used then
+                 * this delay will not be necessary as FreeRTOS_recv() will place the RTOS task
+                 * into the Blocked state anyway. */
+                vTaskDelay( xShortDelay );
+
+                /* Limit the number of FreeRTOS_recv loops to avoid infinite loop. */
+                if( ++xShutdownLoopCount >= mqttexampleMAX_SOCKET_SHUTDOWN_LOOPS )
+                {
+                    break;
+                }
+            }
+
+            /* The socket has shut down and is safe to close. */
+            SOCKETS_Close( xSocket );
+        }
+    }
+}
+/*-----------------------------------------------------------*/
+
+static int32_t prvTransportRecv( const NetworkContext_t * pxContext,
+                                 void * pvBuffer,
+                                 size_t xBytesToRecv )
+{
+    int32_t lResult;
+
+    configASSERT( pxContext != NULL );
+
+    /* Receive upto xBytesToRecv from network. */
+    lResult = ( int32_t ) SOCKETS_Recv( ( Socket_t ) pxContext->xTcpSocket,
+                                        pvBuffer,
+                                        xBytesToRecv,
+                                        0 );
+
+    return lResult;
+}
+/*-----------------------------------------------------------*/
+
+static uint16_t prvGetNextPacketIdentifier()
+{
+    static uint16_t usPacketId = 0;
+
+    usPacketId++;
+
+    /* Since 0 is invalid packet identifier value,
+     * take care of it when it rolls over */
+    if( usPacketId == 0 )
+    {
+        usPacketId = 1;
+    }
+
+    return usPacketId;
+}
+/*-----------------------------------------------------------*/
+
+static Socket_t prvCreateTCPConnectionToBroker( void )
```

Comment:
(Suggestion) I think that the signature would be clearer if the socket is returned as an output parameter, instead of using the `Socket_t` both the as type to represent success/failure and the socket handle for the network connection..
```
static BaseType_t prvCreateTCPConnectionToBroker( Socket_t * pSocket );
```

```
static BaseType_t prvConnectToServerWithBackoffRetries( Socket_t * pSocket )
```